lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -253,23 +255,179 @@ struct BinaryOpLowering : public OpConversionPattern<BinOp> {
     return success();
   }
 };
+
+template <class TOp>
+struct ArgResOpConversion : public OpConversionPattern<TOp> {
+  // Generic pattern which replaces an op by one of the same type, but with
+  // converted operands and result types.
+  using OpConversionPattern<TOp>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<TOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(TOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Use the generic builder to allow this pattern to apply to all ops.
+    llvm::SmallVector<Type, 4> convResTypes;
+    if (failed(this->getTypeConverter()->convertTypes(op->getResultTypes(),
+                                                      convResTypes)))
+      return failure();
+    rewriter.replaceOpWithNewOp<TOp>(op, convResTypes, adaptor.getOperands(),
+                                     op->getAttrs());
+    return success();
+  }
+};
+
+/// A helper type converter class that automatically populates the relevant
+/// materializations and type conversions for converting HWArith to HW.
+class HWArithToHWTypeConverter : public TypeConverter {
+public:
+  HWArithToHWTypeConverter();
+
+  // A function which recursively converts any integer type with signedness
+  // semantics to a signless counterpart.
+  Type removeSignedness(Type type) {
+    auto it = conversionCache.find(type);
+    if (it != conversionCache.end())
+      return it->second.type;
+
+    auto convertedType =
+        llvm::TypeSwitch<Type, Type>(type)
+            .Case<IntegerType>([](auto type) {
+              if (type.isSignless())
+                return type;
+              return IntegerType::get(
+                  type.getContext(), type.getIntOrFloatBitWidth(),
+                  IntegerType::SignednessSemantics::Signless);
```

Comment:
Nit: could be simplified to `IntegerType::get(type.getContext(), type.getWidth())`